File: SimpleMatchIcons/app/src/main/java/com/example/mjjohnso/simplematch/MainActivity.java
package com.example.mjjohnso.simplematch;

import android.app.Activity;
import android.app.AlertDialog;
import android.app.Dialog;
import android.app.DialogFragment;
import android.content.DialogInterface;
import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import static android.content.ContentValues.TAG;

public class MainActivity extends Activity {

    private static final String SCORE = "SCORE";
    private static final String LAST = "LAST";
    private static final String MATCH = "MATCH";
    private static final String NUMS = "NUMS";
    private static final String TURNED = "TURNED";

    private TextView mDone;
    private Button mRestart;
    private int mTileValues[]=new int[16];
    private boolean mTurned[] = new boolean[16];
    private int mScore=0;
    private int mNumMatched=0;
    private int mLastTileIndex =-1;
    private ImageView mTiles[]=new ImageView[16];
    private final int[] mTileIds ={
            R.id.button11,R.id.button12,R.id.button13,R.id.button14,
            R.id.button21,R.id.button22,R.id.button23,R.id.button24,
            R.id.button31,R.id.button32,R.id.button33,R.id.button34,
            R.id.button41,R.id.button42,R.id.button43,R.id.button44
    };

    private int[] mDrawables={
      R.drawable.ic_attachment_black_24dp,
      R.drawable.ic_audiotrack_black_24dp,
      R.drawable.ic_brightness_5_black_24dp,
      R.drawable.ic_brush_black_24dp,
      R.drawable.ic_build_black_24dp,
      R.drawable.ic_flight_black_24dp,
      R.drawable.ic_spa_black_24dp,
      R.drawable.ic_weekend_black_24dp,
    };

    private void setTile(final int i, final int s) {
        final int from;
        final int to;
        if (s == -1) {
            from = 0;
            to = 180;
            mTurned[i]=false;
        } else {
            from = 180;
            to = 0;
            mTurned[i]=true;
        }
        mTiles[i].setRotationY(from);

        mTiles[i].animate().rotationY((from+to) / 2).setDuration(100).withEndAction(() -> {
                if(s!=-1)
                    mTiles[i].setImageResource(mDrawables[s]);
                else
                    mTiles[i].setImageDrawable(null);
                mTiles[i].animate().rotationY(to).setDuration(100).withEndAction(() -> mTiles[i].setRotationY(0));
        });

    }
    @Override
    protected void onSaveInstanceState(Bundle savedInstanceState) {
        savedInstanceState.putInt(SCORE, mScore);
        savedInstanceState.putInt(LAST, mLastTileIndex);
        savedInstanceState.putInt(MATCH, mNumMatched);
        savedInstanceState.putIntArray(NUMS, mTileValues);
        savedInstanceState.putBooleanArray(TURNED,mTurned);
        super.onSaveInstanceState(savedInstanceState);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        mDone=findViewById(R.id.done);
        mRestart=findViewById(R.id.restart);
        mRestart.setOnClickListener(view -> init());
        for(int i=0;i<16;i++) {
            final int v=i;
            mTiles[i]=findViewById(mTileIds[i]);
            mTiles[i].setOnClickListener(view -> {
                if(mTurned[v])
                    return;
                mScore++;
                setTile(v, mTileValues[v]);
                if(mLastTileIndex ==-1) {
                    mLastTileIndex =v;
                } else {
                    if(mTileValues[mLastTileIndex]== mTileValues[v]) {
                        mNumMatched++;
                        mLastTileIndex =-1;
                    } else {
                        setTile(mLastTileIndex,-1);
                        mLastTileIndex =v;
                    }
                }
                showscore();
            });
        }
        if(savedInstanceState!=null) {
            mScore=savedInstanceState.getInt(SCORE,0);
            mLastTileIndex =savedInstanceState.getInt(LAST,-1);
            mNumMatched=savedInstanceState.getInt(MATCH,0);
            mTileValues =savedInstanceState.getIntArray(NUMS);
            mTurned=savedInstanceState.getBooleanArray(TURNED);
            for(int i=0;i<16;i++)
                if(mTurned[i]) {
                    setTile(i, mTileValues[i]);
                }
        }

        if(mScore==0)
            init();
        showscore();
    }

    private void showscore() {
        if(mNumMatched==8) {
            mDone.setText("Complete:"+mScore);
        } else {
            mDone.setText("Score:"+mScore);
        }
    }
    private void init() {
        mNumMatched=0;
        mScore=0;
        mLastTileIndex =-1;
        for (int i=0;i<16;i++) {
            mTiles[i].setImageDrawable(null);
            mTileValues[i]=-1;
            mTurned[i]=false;
        }
        for(int i=0;i<8;i++) {
            int x;
            for(int j=0;j<2;j++) {
                do {
                    x=(int)(Math.random()*16);
                } while(mTileValues[x]!=-1);
                mTileValues[x]=i;
            }
        }
        showscore();
    }
}

File: SimpleMatchIcons/app/src/main/java/com/example/mjjohnso/simplematch/MainActivity.kt
package com.example.mjjohnso.simplematch

import android.os.Bundle
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.animation.core.Spring.DampingRatioMediumBouncy
import androidx.compose.animation.core.Spring.StiffnessMediumLow
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.spring
import androidx.compose.foundation.Image
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Button
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import com.example.mjjohnso.simplematch.ui.theme.AppTheme

const val ROWS=4
const val COLS=4
const val SCORE = "SCORE"
const val LAST = "LAST"
const val MATCH = "MATCH"
const val NUMS = "NUMS"
const val TURNED = "TURNED"

class MainActivity : ComponentActivity() {

    private var score = 0
    private var numMatched = 0
    private var lastButtonIndex = -1

    private var tileValues = Array(ROWS * COLS) {0}
    private var scoreState = mutableStateOf("")
    private var turned = Array(ROWS * COLS) {mutableStateOf(false)}

    private val drawables= intArrayOf(
      R.drawable.ic_attachment_black_24dp,
      R.drawable.ic_audiotrack_black_24dp,
      R.drawable.ic_brightness_5_black_24dp,
      R.drawable.ic_brush_black_24dp,
      R.drawable.ic_build_black_24dp,
      R.drawable.ic_flight_black_24dp,
      R.drawable.ic_spa_black_24dp,
      R.drawable.ic_weekend_black_24dp,
    )

    override fun onSaveInstanceState(outState: Bundle) {
        outState.putInt(SCORE, score)
        outState.putInt(LAST, lastButtonIndex)
        outState.putInt(MATCH, numMatched)
        outState.putIntArray(NUMS, tileValues.toIntArray())
        outState.putBooleanArray(TURNED,turned.map{it.value}.toBooleanArray())
        super.onSaveInstanceState(outState)
    }

    private fun buttonClick(index: Int) {
        if (!turned[index].value) {
            score++
            turned[index].value = true
            if (lastButtonIndex == -1) {
                lastButtonIndex = index
            } else {
                if (tileValues[lastButtonIndex] == tileValues[index]) {
                    numMatched++
                    lastButtonIndex = -1
                } else {
                    val bi=lastButtonIndex
                    lastButtonIndex = index
                    turned[bi].value = false
                }
            }
        }
        showscore()
    }

    @Composable
    fun TurningButton( index: Int, modifier: Modifier = Modifier) {
        val angle by animateFloatAsState(
            targetValue = if (turned[index].value) 0f else 180f,
            animationSpec =  spring( stiffness = StiffnessMediumLow,
                dampingRatio = DampingRatioMediumBouncy)
        )
        Button( shape = RoundedCornerShape(4.dp),
            onClick= { buttonClick(index)},
            modifier = modifier.fillMaxSize()
                .padding(4.dp)
                .graphicsLayer { rotationY=angle },
            contentPadding=PaddingValues(0.dp)) {
            if(angle<=90f && turned[index].value) {
                Image(painter = painterResource(id =  drawables[tileValues[index]]),
                    contentDescription = "",
                    modifier.fillMaxWidth().
                    fillMaxHeight(),
                    contentScale=ContentScale.Fit)
            }
        }
    }

    @Composable
    fun MatchGame(modifier: Modifier = Modifier) {
        Column(modifier = modifier) {
            Row {
                Text(text = scoreState.value, modifier = Modifier.weight(2f).align(Alignment.CenterVertically))
                Button(onClick = { init() }, modifier = Modifier.weight(1f)) {Text("Restart")}
            }
            for (i in 0..<ROWS) {
                Row(modifier = Modifier.weight(1f)) {
                    for (j in 0..<COLS) {
                        val index = i * COLS + j
                        TurningButton(index, Modifier.weight(1f))
                    }
                }
            }
        }
    }
    @Preview(showBackground = true, showSystemUi = true)
    @Composable
    fun ComposablePreview() {
        init()
        AppTheme {
            Scaffold(modifier = Modifier.fillMaxSize()) { innerPadding ->
                MatchGame(modifier = Modifier.padding(innerPadding))
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContent {
            AppTheme {
                Scaffold(modifier = Modifier.fillMaxSize()) { innerPadding ->
                    MatchGame(modifier = Modifier.padding(innerPadding))
                }
            }
        }

        if (savedInstanceState != null) {
            score = savedInstanceState.getInt(SCORE, 0)
            lastButtonIndex = savedInstanceState.getInt(LAST, -1)
            numMatched = savedInstanceState.getInt(MATCH, 0)
            tileValues = savedInstanceState.getIntArray(NUMS)?.map{ it }?.toTypedArray()!!
            turned = savedInstanceState.getBooleanArray(TURNED)?.map{ mutableStateOf(it) }?.toTypedArray()!!
        }
        if (score == 0) init()
        showscore()
    }

    private fun showscore() {
        if (numMatched == ((ROWS * COLS) / 2) ) {
            scoreState.value = "Complete:$score"
        } else {
            scoreState.value = "Score:$score"
        }
    }

    private fun init() {
        numMatched = 0
        score = 0
        lastButtonIndex = -1
        for (i in 0..< ROWS * COLS) {
            tileValues[i] = -1
            turned[i].value = false
        }
        for (i in 0..<(ROWS * COLS)/2) {
            var x: Int
            (0..1).forEach { _ ->
                do {
                    x = (Math.random() * (ROWS * COLS)).toInt()
                } while ( tileValues[x]!=-1)
                tileValues[x] =  i
            }
        }
        showscore()
    }
}
